### ISSUE TITLE
[autodev][design] HF transformers setup migration — design + experiments before any code change

### ISSUE THREAD

**Issue by penguinwu** [penguinwu/oss-model-graph-break-corpus#100]
## Goal (DESIGN ONLY — no migration yet)

Decide whether to migrate the HF transformers create_model path (currently ~3000 lines of heuristic in `sweep/worker.py`) to the same authoritative-fixture pattern we just shipped for diffusers (`_load_diffusers_test_fixture`, commit f513c5d). This issue captures the design and the experiments we'd run BEFORE deciding.

Per Peng (2026-04-29): "any migration is an overhead and takes our momentum away from the actual analysis work" — so the bar to migrate is high. This is design phase, not implementation.

## Background — what we have today

`sweep/worker.py:create_hf_model` (~3000 lines including helpers + per-model patches):
- `_create_config(model_name, config_cls)` — calls `config_cls()` with defaults
- `_fix_config(model_name, config)` — heuristically reduces dims (cap layers, hidden_size, etc.)
- `_detect_hf_input_type()` — infers modality from class name (text/vision/multimodal/audio/seq2seq/...)
- ~12 modality-branch input synthesizers + dozens of per-model special cases (EdgeTam, PerceptionLM, Qwen3OmniMoeTalker, KyutaiSpeechToText, PI0, Zamba2, ...)

Result on PT 2.13: 1081 full_graph + 270 graph_break out of 1482 HF work items = 91% useful signal. Works, but accumulates technical debt with each new model release.

## What HF officially has

Each HF model has `tests/models/<model>/test_modeling_<model>.py` containing:
- A `<Model>ModelTester` class with hand-tuned small dims (e.g. BART: vocab=99, hidden=16, layers=2, heads=4, ffn=4)
- `prepare_config_and_inputs_for_common()` returning (config, inputs_dict)
- `all_model_classes` tuple declaring which model classes the tester covers
- Helper functions `ids_tensor`, `floats_tensor` from `test_modeling_common`

**Coverage measurement (verified)**:
- HF tester `all_model_classes` covers 1085 unique class names across 460 model dirs
- Our `enumerate_hf()` returns 740 model specs
- *Overlap: 509 models* (~69% of our enum has an HF tester fixture)
- 231 models in our enum NOT in HF tests (likely newer models or test-naming mismatches)
- 576 models in HF tests NOT in our enum (mostly AutoModel variants + classes we filter)

### Q1 — testing size: HF official vs ours

HF tester configs are tiny (BART: vocab=99 / hidden=16 / layers=2). Ours uses `_fix_config` which caps via heuristics — typically larger. Smaller HF configs mean faster create + faster compile, but may miss real bugs that only surface at production-ish dim. Need head-to-head comparison.

**Experiment**: pick 20 representative models from the overlap. For each, construct via HF tester config AND via our `_create_config + _fix_config`. Record (a) param count, (b) create_time_s, (c) compile_time_s. Decide: are HF configs "too small" to be useful, or just-right?

### Q2 — coverage: HF official vs ours

We have 740 enum specs; HF has fixtures for ~509 of them. The 231 in our enum without HF fixtures need backup (heuristic fallback OR explicit skip).

**Experiment**: run the HF fixture-loader on all 509 overlap models. Record per model: did fixture load? did construct succeed? did forward succeed? Compare with our heuristic path's pass/fail for the same models. Output: a coverage matrix.

### Q3 — multi-input value for skill discovery (Peng's insight)

Even if HF setup is "better", the skill project NEEDS multiple inputs per (model, mode) to validate that a skill-driven model rewrite holds across input variation. So we shouldn't replace ours with HF's — we should HAVE BOTH and run sweeps with both.

**Design implication**: don't think "migrate", think "add a second input source". The schema gets a `setup_source: hf-fixture | corpus-heuristic | ...` field; sweeps can run with one or both; results tagged accordingly. Skill validation requires running with both and comparing.

**Experiment**: pick 5 near-fullgraph models from the existing skill discovery work (#92 BUILD_STRING family, etc.). Run them with both setups. Compare graph break shapes — same patterns? different shapes that strengthen the skill catalog?

### Q4 — HF official torch.compile config

**HF DOES officially test torch.compile**. Found in `tests/test_modeling_common.py`:
- `@pytest.mark.torch_compile_test` marker
- `test_sdpa_can_compile_dynamic` — runs `torch.compile(model, dynamic=True)`, requires `_supports_sdpa=True` on model class
- `test_flash_attn_2_can_compile_with_attention_mask_None_without_graph_break` — *explicitly tests no graph breaks!* This is exactly what our sweep tests.
- `_torch_compile_train_cls` attribute — per-model compile-train test class

**Implications**:
- HF maintains a list of which models are "compile-supported" via `_supports_sdpa`
- Their no-graph-break test is the most directly comparable signal to our sweep
- Their `dynamic=True` setting is the canonical compile config

**Experiment**: enumerate HF's torch.compile-tested models. Cross-reference with our HF sweep results. Are the models HF officially tests for compile = the models that get full_graph in our sweep? If not, why the divergence? Are there models we test that HF doesn't, and vice versa?

## Decision criteria (after experiments)

The migration decision should answer:
1. **Coverage**: Does HF fixture path cover ≥90% of our useful models cleanly?
2. **Quality**: Are HF configs producing same-or-better signal (compile success rate, graph break shape diversity)?
3. **Maintenance cost**: How much per-model patching does HF's path also need? (Hopefully less, but verify.)
4. **Multi-input value**: Even if HF wins, do we keep ours as a parallel source for skill validation?

If criteria 1+2 met: phased migration (start with overlap models, retire heuristic in stages).
If criteria 1+2 NOT met: keep heuristic, treat HF fixtures as a NEW additional input source (per Q3).

## Out of scope

- Implementing the migration
- Per-model patches that are working today
- Modifying `enumerate_hf` (this is purely about create_model / inputs)

## Suggested execution order

1. Q4 — find HF compile config + cross-reference with our sweep results (cheapest, may immediately tell us a lot)
2. Q1 — head-to-head config size on 20 models (fast)
3. Q2 — full coverage matrix on 509 overlap models (runs overnight)
4. Q3 — multi-input value experiment (small, depends on having both setups working side-by-side)

Deliverable: a brief design memo answering all four with data, then Peng decides whether to migrate.

**Comment by penguinwu** [comment]
## Q4 update — torch.compile scope dig (2026-04-29 follow-up)

Per Peng's question about scope: dug into HF's torch.compile usage. Three patterns total:

### Pattern 1 — top-level model compile (HF's main test path)

`tests/test_modeling_common.py` has two compile tests, both wrap the WHOLE model:
- `test_flash_attn_2_can_compile_with_attention_mask_None_without_graph_break`: `torch.compile(model, fullgraph=True)` — strict, no graph breaks. Same scope as our sweep.
- `test_sdpa_can_compile_dynamic`: `torch.compile(model, dynamic=True)` — tolerates breaks, tests dynamic shapes.

The "sdpa" / "flash_attn_2" in test names refers to which **attention backend** (`attn_implementation=...`) is loaded — not the compile scope. Compile scope = top-level in both.

### Pattern 2 — forward-method-only compile (HF docs, not in common tests)

Documented in `docs/source/en/model_doc/modernbert.md` and `experts_interface.md`:

```python
model.forward = torch.compile(model.forward, fullgraph=True)
model.forward = torch.compile(model.forward, mode="max-autotune-no-cudagraphs")
```

Compiles ONLY the `forward` method. Sidesteps generation loop, cache management, and other per-call init/cleanup that's hostile to compile but not part of the actual forward pass.

### Pattern 3 — `TrainingArguments(torch_compile=True)` (training, not what we test)

### Implications for the design

- We use Pattern 1 (top-level + fullgraph + eager backend). Aligned with HF's strict-test path.
- Pattern 2 (forward-only) is an INTERESTING ALTERNATIVE INPUT SOURCE for the skill discovery project — same model, different compile scope produces different break categories. Worth deferring as a follow-up experiment IF skill discovery wants more variation per-model.
- The `_supports_sdpa` / `_supports_flash_attn` registry flags are still HF's "compile-readiness per model" registry — cross-referencing our sweep results against these flags remains the highest-value Q4 experiment.

### Updated Q4 experiment list

1. (highest value) Cross-reference our HF sweep results against HF's `_supports_sdpa` and `_supports_flash_attn` registry. Models that HF says "compile-ready" but our sweep says "graph_break" → real divergence worth digging into.
2. (deferred) Compare Pattern 1 (top-level) vs Pattern 2 (forward-only) compile on a sample of generation models — does forward-only avoid breaks that top-level hits?
3. (NOT pursuing per Peng) Per-attention-backend testing (sdpa vs flash_attn_2 vs default) — not a priority.
